File: 2018/5/reactions2.nim
import future
import strutils
import tables

var data = readFile("input.txt").strip()

let backup = data

var i = 0
var stack: seq[char]
stack = @[]

while i<len(data):
    if len(stack)==0:
      stack.add(char(data[i]))
    elif (stack[^1].toUpperAscii()==data[i].toUpperAscii()) and not(stack[^1]==data[i]):
      discard stack.pop()
    else:
      stack.add(data[i])
    i += 1
echo(len(stack))

var results = initTable[char,int]()
let letters = {'a'..'z'}
              
for letter in letters:
  data = backup
  data = data.replace($letter)
  data = data.replace($letter.toUpperAscii())
  i = 0
  stack = @[]
  while i<len(data):
    if len(stack)==0:
      stack.add(data[i])
    elif (stack[^1].toUpperAscii()==data[i].toUpperAscii()) and not(stack[^1]==data[i]):
      discard stack.pop()
    else:
      stack.add(data[i])
    i += 1
  results[letter] = len(stack)

var m = len(backup)
for i,v in results:
  if v<m:
    m = v
echo m

File: 2018/5/reactions2.py
from string import ascii_lowercase
from operator import itemgetter

data = [i.strip() for i in open("input.txt").readlines()][0]

copy = data

i = 0
stack = []
while i<len(data):
    if len(stack)==0:
        stack.append(data[i])
    elif (stack[-1].upper()==data[i].upper()) and not(stack[-1]==data[i]):
        stack.pop()
    else:
        stack.append(data[i])
    i += 1
print(len(stack))

results = []
for letter in ascii_lowercase:
    data = copy.replace(letter,"").replace(letter.upper(),"")
    i = 0
    stack = []
    while i<len(data):
        if len(stack)==0:
            stack.append(data[i])
        elif (stack[-1].upper()==data[i].upper()) and not(stack[-1]==data[i]):
            stack.pop()
        else:
            stack.append(data[i])
        i += 1
    results.append([letter,len(stack)])

ranked = sorted(results, key=itemgetter(1))

print(ranked[0][1])
